Principle: Continuously maintain and evolve (#505)
UKHomeOffice/engineering-guidance-and-standards: docs/principles/continuously-maintain-and-evolve.md

---
layout: principle
order: 1
title: Continuously maintain and evolve
date: 2025-01-17
tags:
  - Ways of working
  - Infrastructure
  - Dependencies
  - Security
  - SRE
related:
  sections:
    - title: Related links
      items:
        - text: Write maintainable, reusable and evolutionary code
          href: /principles/write-maintainable-reusable-and-evolutionary-code/
        - text: Well managed code

>           href: /principles/well-managed-code/

Reviewer: Could add this video as a related link: https://www.youtube.com/watch?v=jjFgyAOIhSg

Reviewer: Or could incorporate points from the video

Author: This is a great talk that builds on Team Topologies, focusing primarily on ways of working, team structures, and organisational culture in software delivery. I think these ideas could be further explored as a separate set of principles or perhaps as a practical pattern, incorporating shared experiences and guidance on how they can be effectively implemented.

Reviewer 2: @dk4g @aaronrussellHO be good to get this conv resolved with adding video or comments or something

Reviewer: I think we're happy to resolve and merge. As BRD guild are iterating and will open a PR with edits. @GraemeSutherlandHO @dk4g

---

Building great products isn't just about writing good code initially - it's about continuously maintaining and evolving them. Regularly update codebases, dependencies, pipelines, infrastructure, and processes to ensure they remain secure, performant, and aligned with current standards. Avoid allowing systems to stagnate by scheduling maintenance and improvements as a regular part of the delivery lifecycle.

---

## Rationale

Outdated systems are harder to support and integrate with modern components, leading to increased technical debt over time. Regular updates are essential to reduce exposure to known vulnerabilities, enhancing the security of the system. Keeping tools and libraries up to date also sustains efficiency by leveraging improved performance and new features. Neglecting regular maintenance often results in costly big bang upgrades, making continuous updates a more cost-effective approach.

---

## Applications and Implications

> - Integrate regular time slots within each sprint or release cycle dedicated to updating dependencies, refactoring code, and maintaining pipelines. This proactive approach ensures that systems remain current and reduces the risk of accumulating technical debt
> - Establish a process to continuously observe and assess updates in the technology ecosystem, such as new library versions or framework releases. Staying informed enables timely adoption of improvements and ensures compatibility with external systems
> - Implement tools that automatically detect outdated dependencies and facilitate testing processes. Automation streamlines maintenance tasks, minimises human error, and accelerates the integration of necessary updates
> - Incorporate planned and measurable maintenance activities into team roadmaps. By formally recognising maintenance as a critical component of the development lifecycle, teams can allocate resources effectively and maintain system integrity over time

Reviewer 3: Some minor changes to make these simpler and be less specific about the implementation details

```suggestion
- Schedule time dedicated to updating dependencies, refactoring code, and maintaining pipelines. This proactive approach ensures that systems remain current and reduces the risk of accumulating technical debt
- Continuously observe and assess updates in the technology ecosystem, such as new library versions or framework releases. Staying informed enables timely adoption of improvements and ensures compatibility with external systems
- Automatically detect outdated dependencies and facilitate testing processes. Automation streamlines maintenance tasks, minimises human error, and accelerates the integration of necessary updates
- Incorporate planned and measurable maintenance activities into team roadmaps. By formally recognising maintenance as a critical component of the development lifecycle, teams can allocate resources effectively and maintain system integrity over time
```

> - Incorporate planned and measurable maintenance activities into team roadmaps. By formally recognising maintenance as a critical component of the development lifecycle, teams can allocate resources effectively and maintain system integrity over time

Reviewer: Consider accessibility when making changes.

Author: Added recommendation to review HO SEGAS regularly
